File: musicShopProject/tentashop/src/tools/types/dataResult.ts
import { Result } from "./results";

export class DataResult<T>{
    constructor(
       public result: Result,
       public data: T | null
    ){}
    
    static fail<T>(error: string): DataResult<T> {
        return new DataResult<T>(Result.fail([error]), null);
    }

    static success<T>(data: T): DataResult<T> {
        return new DataResult<T>(Result.success(), data);
    }
}

File: musicShopProject/Tools/Types/DataResult.cs
namespace musicShopProject.Tools.Types;

public class DataResult<T>
{
    public Result Result { get; }
    public T? Data { get; }

    public DataResult(Result result, T? data)
    {
        Result = result;
        Data = data;
    }

    public static DataResult<T?> Fail(String error)
    {
        return new DataResult<T?>(Result.Fail(error), default(T?));
    }

    public static DataResult<T> Success(T data)
    {
        return new DataResult<T>(Result.Success(), data);
    }
}
